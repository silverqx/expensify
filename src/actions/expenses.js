import database from '../firebase/firebase'
import * as types from '../constants/expenseActionTypes'

export const addExpense = (expense) => ({
    type: types.ADD_EXPENSE,
    expense
})

export const startAddExpense = ({
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
} = {}) => {
    return (dispatch) => {
        const expense = {
            description,
            note,
            amount,
            createdAt
        }

        return database.ref('expenses').push(expense).then((snapshot) => {
            dispatch(addExpense({
                id: snapshot.key,
                ...expense
            }))
        })
    }
}

export const editExpense = (id, updates) => ({
    type: types.EDIT_EXPENSE,
    id,
    updates
})

export const removeExpense = (id) => ({
    type: types.REMOVE_EXPENSE,
    id
})
