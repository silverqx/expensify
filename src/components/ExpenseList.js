import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'

import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = ({ history, expenses }) => (
    <div className="content-container">
        <div className="list">
            <div className="list__header">
                <div className="show-on-desktop">Expense</div>
                <div className="show-on-desktop">Amount</div>
                <div className="show-on-mobile">Expenses</div>
            </div>

            <div className="list__content">
                {
                    expenses.length === 0 ?
                        <div className="list__content--empty">No Expenses</div> :
                        expenses.map((expense) => (
                            <ExpenseListItem
                                key={expense.id}
                                {...expense}
                            />
                        ))
                }
            </div>
        </div>
    </div>
)

const mapStateToProps = ({ expenses, filters }) => ({
    expenses: getVisibleExpenses(expenses, filters)
})

export default connect(mapStateToProps)(ExpenseList)
