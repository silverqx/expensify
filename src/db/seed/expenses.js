import moment from 'moment'

const expensesSeed = [{
    description: 'Water Bill',
    note: 'Water Bill Note',
    amount: 166.60,
    createdAt: moment('2019-04-19').valueOf()
}, {
    description: 'Gas Bill',
    note: 'Gas Bill Note',
    amount: 5.20,
    createdAt: moment('2019-03-17').valueOf()
}, {
    description: 'Internet Bill',
    note: 'Internet Bill Note',
    amount: 50.20,
    createdAt: moment('2019-04-09').valueOf()
}, {
    description: 'Telekom Bill',
    note: 'Telekom Bill Note',
    amount: 10.20,
    createdAt: moment('2019-01-10').valueOf()
}]

export default expensesSeed
