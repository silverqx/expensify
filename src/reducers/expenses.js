import * as types from '../constants/expenseActionTypes'

const expensesReducerInitialState = []

const expensesReducer = (state = expensesReducerInitialState, action) => {
    switch (action.type) {
        case types.ADD_EXPENSE:
            return [
                ...state,
                action.expense
            ]
        case types.EDIT_EXPENSE:
            return state.map((expense) => {
                if (expense.id !== action.id)
                    return expense

                return {
                    ...expense,
                    ...action.updates
                }
            })
        case types.REMOVE_EXPENSE:
            return state.filter((expense) => expense.id !== action.id)
        case types.SET_EXPENSES:
            return action.expenses
        default:
            return state
    }
}

export default expensesReducer
