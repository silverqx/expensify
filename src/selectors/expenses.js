import moment from 'moment'

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        const createdAt = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        const sortByMap = {
            date: 'createdAt',
            amount: 'amount'
        }
        const sortColumn = sortByMap[sortBy]

        if (a[sortColumn] < b[sortColumn])
            return -1
        if (a[sortColumn] > b[sortColumn])
            return 1

        return 0
    })
}

export default getVisibleExpenses
