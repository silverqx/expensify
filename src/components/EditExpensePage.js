import React, { Component } from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'

import { editExpense } from '../actions/expenses'

export class EditExpensePage extends Component {
    onSubmit = (updatedExpense) => {
        this.props.editExpense(
            this.props.match.params.id,
            updatedExpense
        )
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

// FIXME really bad solution, should be done in router? And how to test this? silver
const findExpense = (state, { history, match }) => {
    const expense = state.expenses.find(
        expense => expense.id === match.params.id
    )

    if (!expense)
        return history.goBack()

    return expense
}

const mapStateToProps = (state, props) => ({
    expense: findExpense(state, props)
})

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updatedExpense) => dispatch(
        editExpense(id, updatedExpense)
    )
})

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage)
