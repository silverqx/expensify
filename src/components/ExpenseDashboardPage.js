import React from 'react'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage = () => (
    <div>
        <h1>Expense Dashboard Page</h1>
        <h3>Expenses List</h3>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage
