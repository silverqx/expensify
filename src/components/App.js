import React, { Component } from 'react'
import { Router, Route, Switch } from "react-router-dom"
import ReactModal from 'react-modal'
import { createBrowserHistory } from 'history'

import ExpenseDashboardPage from './ExpenseDashboardPage'
import AddExpensePage from './AddExpensePage'
import EditExpensePage from './EditExpensePage'
import LoginPage from './LoginPage'
import NotFoundPage from './NotFoundPage'

import ModalOption from './ModalOption'
import PrivateRoute from '../routers/PrivateRoute'
import PublicRoute from '../routers/PublicRoute'

// import { clearInputValue } from '../utils/utils'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

// Make sure to bind modal to your appElement
// (http://reactcommunity.org/react-modal/accessibility/)
ReactModal.setAppElement('#app')

export const history = createBrowserHistory()

export default class ExpensifyApp extends Component {
    static defaultProps = {
        title: 'Expensify'
    }

    state = {
        error: null,
        // TODO implement subtitle silver
        subtitle: 'Xxx',
    }

    componentDidMount() {
    }

    componentDidUpdate(prevProps, prevState) {
    }

    /**
     * Update Error state.
     *
     * @param {string} error - Error message.
     *
     * @returns {?boolean} True every time.
     */
    handleError = (error) => {
        this.setState(() => ({ error }))

        return true
    }

    /**
     * Clear Error state.
     *
     * @returns {?boolean} False every time.
     */
    clearError = () => {
        this.setState(() => ({ error: null}))

        return false
    }

    render() {
        return (
            <Router history={history}>
                <Switch>
                    <PublicRoute path="/" exact component={LoginPage} />
                    <PrivateRoute path="/dashboard" component={ExpenseDashboardPage} />
                    <PrivateRoute path="/create" component={AddExpensePage} />
                    <PrivateRoute path="/edit/:id" component={EditExpensePage} />
                    <Route component={NotFoundPage} />
                </Switch>
                <ModalOption
                    selectedOption={this.state.selectedOption}
                    handleClearSelectedOption={this.handleClearSelectedOption}
                />
            </Router>
        )
    }
}
