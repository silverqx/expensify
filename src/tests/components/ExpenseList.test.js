import React from 'react'
import { shallow } from 'enzyme'

import { ExpenseList } from '../../components/ExpenseList'

import { expensesWithIds as expenses } from '../fixtures/expenses'

describe('ExpenseList component', () => {
    test('should render ExpenseList with expenses', () => {
        const wrapper = shallow(<ExpenseList expenses={expenses} />)

        expect(wrapper).toMatchSnapshot()
    })

    test('should render ExpenseList with empty expenses', () => {
        const wrapper = shallow(<ExpenseList expenses={[]} />)

        expect(wrapper).toMatchSnapshot()
    })
})
