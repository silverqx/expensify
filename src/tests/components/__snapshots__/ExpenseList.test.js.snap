// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList component should render ExpenseList with empty expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list"
  >
    <div
      className="list__header"
    >
      <div
        className="show-on-desktop"
      >
        Expense
      </div>
      <div
        className="show-on-desktop"
      >
        Amount
      </div>
      <div
        className="show-on-mobile"
      >
        Expenses
      </div>
    </div>
    <div
      className="list__content"
    >
      <div
        className="list__content--empty"
      >
        No Expenses
      </div>
    </div>
  </div>
</div>
`;

exports[`ExpenseList component should render ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list"
  >
    <div
      className="list__header"
    >
      <div
        className="show-on-desktop"
      >
        Expense
      </div>
      <div
        className="show-on-desktop"
      >
        Amount
      </div>
      <div
        className="show-on-mobile"
      >
        Expenses
      </div>
    </div>
    <div
      className="list__content"
    >
      <Connect(ExpenseListItem)
        amount={166.6}
        createdAt={1555624800000}
        description="Water Bill"
        id="1"
        key="1"
        note="Water Bill Note"
      />
      <Connect(ExpenseListItem)
        amount={5.2}
        createdAt={1552777200000}
        description="Gas Bill"
        id="2"
        key="2"
        note="Gas Bill Note"
      />
      <Connect(ExpenseListItem)
        amount={50.2}
        createdAt={1554760800000}
        description="Internet Bill"
        id="3"
        key="3"
        note="Internet Bill Note"
      />
      <Connect(ExpenseListItem)
        amount={10.2}
        createdAt={1547074800000}
        description="Telekom Bill"
        id="4"
        key="4"
        note="Telekom Bill Note"
      />
    </div>
  </div>
</div>
`;
